### PR TITLE
WiP: Omhiredis now reconnects to redis after failure

### PR DESCRIPTION
**Pull request by bobthemighty**
* We check for the connection state when receiving replies
  so that we don't try to free(NULL)
* We now close the redis connection if we fail to send logs, which
  causes us to reconnect in tryResume
* Cleaned up the freeReply code in endTransaction and dropped the
  replies array since it isn't actually used.

### CODE REVIEW

**Comment by bobthemighty** [comment]
@rgerhards Fixed the segfault when redis is unavailable, and fixed the issue where omhiredis would not reconnect to a redis server after its connection had failed.

What's confusing me at the moment is that messages that were queued when the worker was suspended don't seem to be forwarded on once redis has come back to life.

Am I missing something really obvious?

**Comment by bobthemighty** [comment]
Yes, I was missing some configuration. Using action.resumeretrycount causes the action to send messages once a redis connection is re-established. I want to do some more error handling, but I think this is close to working. It certainly works better than the old version, which segfaulted as soon as redis went away.

**Comment by pemontto** [comment]
@bobthemighty I just ran a quick test of this in our dev environment and can't seem to get recover messages pushed while Redis is down.

```bash
[root@dev ~]# logger -t test 1
[root@dev ~]# logger -t test 2
[root@dev ~]# logger -t test 3
[root@dev ~]# service redis stop
Stopping redis-server:                                     [  OK  ]
[root@dev ~]# logger -t test 4
[root@dev ~]# logger -t test 5
[root@dev ~]# logger -t test 6
[root@dev ~]# logger -t test 7
[root@dev ~]# service redis start
Starting redis-server:                                     [  OK  ]
[root@dev ~]# logger -t test 8
[root@dev ~]# logger -t test 9
[root@dev ~]# logger -t test 10
[root@dev ~]# logger -t test 11
```

```bash
redis 127.0.0.1:6379> lrange rsyslog:test 0 -1
1) " 11"
2) " 10"
3) " 9"
4) " 8"
5) " 3"
6) " 2"
7) " 1"
```

As you suggested I added resumeRetryCount to my config.
```
        action.resumeRetryCount="-1"
        action.resumeInterval="10"
```

Do you have a full config example that is working for you?

**Comment by bobthemighty** [comment]
HI @pemontto 

I've whipped up a quick gist: https://gist.github.com/bobthemighty/5788f72a1131dd05c3ad005ce5b7e481

It's important to check that you're using an explicit queue; if rsyslog is running in firehose mode (the default) then you will drop messages while redis is unavailable.

**Comment by mzahorik** [comment]
I second this commit.  I'm looking to use this module in our new ELK stack, and reconnect / not crashing upon redis failure, is critical.  Also, I'm trying to get the debian maintainer to include the module, but he asks that it be "production ready".

**Comment by bobthemighty** [comment]
We are now running this in prod and it *seems to be working. I'm happy to
take off the WIP label if it's working for others.

 -- Bob

**Comment by mzahorik** [comment]
Let me merge this into to a vanilla 8.16.0 package, and I'll give it a try.   It'll probably be a few days/a week before I'll have a chance to test fully.

**Comment by mostolog** [comment]
Hi.

We are also doing some tests and experiencing redis segfaults/crashes that seem to happen when  connection is closed.

Even worse: if setting a docker cluster each request is redirected to a different node, so omhiredis is closing-opening connections frecuently => crashing.

Is there anything we can test/help to get this issue fixed?

Regards

**Comment by bobthemighty** [comment]
@mostolog you can try building rsyslog from our fork, or just merging this pr into a branch. If it doesn't help with your Docker problem I'd like to know so we can try something else.

Once I get confirmation from a few people ( @mzahorik @pemontto) that it works for them, I'll pester @rgerhards to merge for the next release.

**Comment by bobthemighty** [comment]
@mostolog you can try building rsyslog from our fork, or just merging this pr into a branch. If it doesn't help with your Docker problem I'd like to know so we can try something else.

Once I get confirmation from a few people ( @mzahorik @pemontto) that it works for them, I'll pester @rgerhards to merge for the next release.

**Comment by mzahorik** [comment]
I got a chance to compile last weekend, and only tested tonight.

It reconnects fine, but I lose logs.  I get:

rsyslogd[internal_messages]: 4 messages lost due to rate-limiting

I set these additional values:

$SystemLogRateLimitInterval 0
$SystemLogRateLimitBurst 0

And here's my action stanza in /etc/rsyslog.d/loghost.conf

action(
  action.resumeretrycount="-1"
  action.resumeinterval="5"
  queue.size="1000000"
  queue.type="LinkedList"
  type="omhiredis"
  mode="publish"
  key="rsyslog_logstash"
  template="json-template"
)

You mentioned "firehose mode", but I couldn't find a reference to that, I assume assigning a explicit queue is done with the queue.type setting.

So in summary, omhiredis seems to be working fine (I'd like to do more testing still) but my issue is rsyslog configuration.

Thanks!

- Matt

**Comment by pemontto** [comment]
@bobthemighty just wanted to give you feedback and say we've been running this for about two weeks now without issue. Massive 👍

**Comment by mostolog** [comment]
@bobthemighty You mean just https://github.com/rsyslog/rsyslog/pull/1109/commits/f1031634aa439e12f4b398b82bd057a0c8359c4e ? I'll try ASAP and give you feedback.
Neither know what's the firehose mode, but as we are looking for 0% loss, it should be disabled (somehow).

In the meantime we have been planning to test rsyslog+omkafka->Kafka+Streams.

Regards

**Comment by bobthemighty** [comment]
@mostolog there's a gist posted earlier in the thread that has an rsyslog
config which uses disk-backed queues for omhiredis. It's not a *great*
config, but I'd start there and see if it works for your use-case.

Alternatively, let me know how Kafka goes for you - I was tempted by that,
but we have existing knowledge and experience with redis, so it was a
simpler option.

**Comment by mostolog** [comment]
@bobthemighty without using rsyslog-buffer-for-non-delivered-messages (we'll try next week) to avoid losing messages, and RELP as protocol, we have been able to send several logs from rsyslog to Redis with your patched file.

We have started a single cluster node without adding slots (ie: clusterdown) and it hasn't failed/crashed as used to do.
Also started several Redis instances (not clustered, but load balanced) and has worked perfectly
Finally we have killed some of them, and even all, and still no crashes.

Next week, we'll try to test how it goes under heavy load and if any message are lost in the way too.

Regards

**Comment by bobthemighty** [comment]
@rgerhards comments in this thread and #1140 suggest that this PR fixes the segfault at least, so I'm comfortable that it's a significant improvement.

 What can I do to move this forward and aim for a merge before the next release? It seems like the original issue is affecting quite a few users.

**Comment by rgerhards** [comment]
@bobthemighty Great work, thanks! 

If possible, could you rebase to the current master branch? That should also make the travis failures go away. I'll see that I merge within the next couple of days.

If the doc needs updates, it would be great if you could add a PR at the rsyslog-doc project.

**Comment by bobthemighty** [comment]
Hi @rgerhards I rebased, after a particularly galling false start, I now have a new build error, but I'm reasonably happy that it's not my doing.

There's no new documentation required, since there's no change to intended functionality - this is just a bug fix to existing behaviour.

**Comment by mostolog** [comment]
Just updating. Using the following configuration, **which if I'm not wrong it's using queues**:

```
action(
    mode="queue"
    name="omhiredis"
    serverport="6379"
    serverpassword="password"
    server="myserver"
    type="omhiredis"
    key="$!myurl_$!myhost"
    action.resumeRetryCount="-1"
    action.reportSuspension="on"
    queue.maxdiskspace="1G"
    queue.type="LinkedList"
    queue.filename="omhiredis.queue"
    queue.SaveOnShutdown="on"
)
```
**RELP->Rsyslog->Redis is working perfectly**

A couple of questions to double-check: **If queue reaches 1GB/full...it won't drop any message, but reject new RELP input, isn't it?**

Now, as you may noticed, I have to find a way to dynamically set _key="$!myurl_$!myhost"_.
Messages will probably be JSON encoded, but setting vars in any way seems to work. 
Probably, as pointed by @radu-gheorghe I'll have to use templates.

@bobthemighty We will probably start testing kafka next moth, so I'll try to keep you updated

**Comment by radu-gheorghe** [comment]
> A couple of questions to double-check: If queue reaches 1GB/full...it won't drop any message, but reject new RELP input, isn't it?
Yep, unless you set some "discard" options (either in the action or in the main queue: http://www.rsyslog.com/doc/master/rainerscript/queue_parameters.html), that's what should happen.
